File: app/src/main/org/runnerup/features/StartTrackerBinding.java
package org.runnerup.features;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import org.runnerup.tracking.Tracker;

/** Manages binding to the GPS {@link Tracker} service from {@link StartFragment}. */
final class StartTrackerBinding {

  interface Callback {
    void onTrackerBound(Tracker tracker);

    void onTrackerUnbound();
  }

  private final Context appContext;
  private final Callback callback;
  private boolean bound;
  private Tracker tracker;

  private final ServiceConnection connection =
      new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName className, IBinder service) {
          tracker = ((Tracker.LocalBinder) service).getService();
          callback.onTrackerBound(tracker);
        }

        @Override
        public void onServiceDisconnected(ComponentName className) {
          if (tracker != null) {
            callback.onTrackerUnbound();
          }
          tracker = null;
        }
      };

  StartTrackerBinding(Context context, Callback callback) {
    this.appContext = context.getApplicationContext();
    this.callback = callback;
  }

  Tracker getTracker() {
    return tracker;
  }

  boolean isBound() {
    return bound;
  }

  void bind() {
    bound =
        appContext.bindService(
            new Intent(appContext, Tracker.class), connection, Context.BIND_AUTO_CREATE);
  }

  void unbind() {
    if (bound) {
      appContext.unbindService(connection);
      bound = false;
    }
    if (tracker != null) {
      callback.onTrackerUnbound();
    }
    tracker = null;
  }
}

File: app/src/main/kotlin/org/runnerup/features/StartTrackerBinding.kt
package org.runnerup.features

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import org.runnerup.tracking.Tracker

/** Manages binding to the GPS [Tracker] service from [StartFragment]. */
internal class StartTrackerBinding(context: Context, private val callback: Callback) {

  interface Callback {
    fun onTrackerBound(tracker: Tracker)

    fun onTrackerUnbound()
  }

  private val appContext = context.applicationContext
  private var bound = false
  var tracker: Tracker? = null
    private set

  private val connection: ServiceConnection =
      object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
          tracker = (service as Tracker.LocalBinder).service
          callback.onTrackerBound(tracker!!)
        }

        override fun onServiceDisconnected(className: ComponentName) {
          if (tracker != null) {
            callback.onTrackerUnbound()
          }
          tracker = null
        }
      }

  fun isBound(): Boolean = bound

  fun bind() {
    bound =
        appContext.bindService(
            Intent(appContext, Tracker::class.java),
            connection,
            Context.BIND_AUTO_CREATE,
        )
  }

  fun unbind() {
    if (bound) {
      appContext.unbindService(connection)
      bound = false
    }
    if (tracker != null) {
      callback.onTrackerUnbound()
    }
    tracker = null
  }
}
